### PR TITLE
KTOR-1268 Fix `readRemaining` lost bytes

### PR DESCRIPTION
**Pull request by e5l**
Fixes [KTOR-1268](https://youtrack.jetbrains.com/issue/KTOR-1268)

### CODE REVIEW

**Review by rsinukov — APPROVED** [review]
LGTM
nit: should we deprecate `joinTo` and remove it in a major release if there are no usages left

**Review by rsinukov — CHANGES_REQUESTED** [review]
sorry, didn't mention it first. 
`DeflaterReadChannelTest.testSmallPieces` was running for 3 hours and timed out

File: ktor-io/jvm/src/io/ktor/utils/io/ByteBufferChannel.kt [hunk 1]
```diff
@@ -1318,14 +1298,19 @@ internal open class ByteBufferChannel(
...
     internal suspend fun copyDirect(src: ByteBufferChannel, limit: Long, joined: JoiningState?): Long {
-        if (limit == 0L) return 0L
+        require(limit > 0) { "Limit should be positive." }
+        println("CD start")
+
         if (src.isClosedForRead) {
             if (joined != null) {
                 check(src.tryCompleteJoining(joined))
             }
+            println("CD return 1: ${state.capacity}")
```
Comment (cy6erGn0m) [inline ktor-io/jvm/src/io/ktor/utils/io/ByteBufferChannel.kt:1308]: println

File: ktor-utils/jvm/test/io/ktor/tests/utils/DeflaterReadChannelTest.kt [hunk 2]
```diff
@@ -50,11 +57,14 @@ class DeflaterReadChannelTest : CoroutineScope {
     @Test
     fun testSmallPieces() {
         val text = "The quick brown fox jumps over the lazy dog"
-        assertEquals(text, asyncOf(text).toInputStream().reader().readText())
+//        assertEquals(text, asyncOf(text).toInputStream().reader().readText())
 
         for (step in 1..text.length) {
-            testReadChannel(text, asyncOf(text))
+            println("Step: $step")
```
Comment (cy6erGn0m) [inline ktor-utils/jvm/test/io/ktor/tests/utils/DeflaterReadChannelTest.kt:63]: println

File: ktor-utils/jvm/test/io/ktor/tests/utils/DeflaterReadChannelTest.kt [hunk 3]
```diff
@@ -100,10 +110,22 @@ class DeflaterReadChannelTest : CoroutineScope {
     private fun testWriteChannel(expected: String, src: ByteReadChannel) {
         val channel = ByteChannel()
         launch {
+            println("Launched")
```
Comment (cy6erGn0m) [inline ktor-utils/jvm/test/io/ktor/tests/utils/DeflaterReadChannelTest.kt:113]: more debug stuff

File: ktor-utils/jvm/test/io/ktor/tests/utils/DeflaterReadChannelTest.kt [hunk 4]
```diff
@@ -100,10 +110,22 @@ class DeflaterReadChannelTest : CoroutineScope {
...
             src.copyAndClose((channel as ByteWriteChannel).deflated())
+            println("Copied")
         }
 
-        val result = channel.toInputStream().ungzip().reader().readText()
-        assertEquals(expected, result)
+        runBlocking {
```
Comment (cy6erGn0m) [inline ktor-utils/jvm/test/io/ktor/tests/utils/DeflaterReadChannelTest.kt:118]: Please revert the test changes

**Review by cy6erGn0m — CHANGES_REQUESTED** [review]

**Review by rsinukov — APPROVED** [review]

**Review by cy6erGn0m — APPROVED** [review]

**Pull request merged**